Validate migration IDs

marvil07 opened the pull request:
**Describe the bug**
`migrate:stop`, `migrate:messages` and `migrate:fields-source` are not handling the case when an invalid migrate name is passed.

**To Reproduce**
Try `vendor/bin/drush migrate:stop unexisting_migration`
Change it for `migrate:messages` and `migrate:fields-source` to get similar results.

**Expected behavior**
An error message is presented.

**Actual behavior**
An unhandled error is reached, triggering a php fatal error.

**Workaround**
No.

### System Configuration
| Q               | A
| --------------- | ---
| Drush version?  | 10.x e436955b5556113dda0af8a3c4225f0b209ed329
| Drupal version? | 9.x
| PHP version     | 7.3
| OS?             | Linux

**Additional information**
Add any other context about the problem here.

Here the full output I get without the extra change on this pull request.

    $ vendor/bin/drush migrate:stop unexisting_migration
     [error]  Error: Call to a member function getStatus() on bool in Drush\Drupal\Commands\core\MigrateRunnerCommands->stop() (line 438 of vendor/drush/drush/src/Drupal/Commands/core/MigrateRunnerCommands.php) #0 [internal function]: Drush\Drupal\Commands\core\MigrateRunnerCommands->stop('unexisting_migr...', Array)
    #1 vendor/consolidation/annotated-command/src/CommandProcessor.php(257): call_user_func_array(Array, Array)
    #2 vendor/consolidation/annotated-command/src/CommandProcessor.php(212): Consolidation\AnnotatedCommand\CommandProcessor->runCommandCallback(Array, Object(Consolidation\AnnotatedCommand\CommandData))
    #3 vendor/consolidation/annotated-command/src/CommandProcessor.php(176): Consolidation\AnnotatedCommand\CommandProcessor->validateRunAndAlter(Array, Array, Object(Consolidation\AnnotatedCommand\CommandData))
    #4 vendor/consolidation/annotated-command/src/AnnotatedCommand.php(311): Consolidation\AnnotatedCommand\CommandProcessor->process(Object(Symfony\Component\Console\Output\ConsoleOutput), Array, Array, Object(Consolidation\AnnotatedCommand\CommandData)) 
    #5 vendor/symfony/console/Command/Command.php(255): Consolidation\AnnotatedCommand\AnnotatedCommand->execute(Object(Drush\Symfony\DrushArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
    #6 vendor/symfony/console/Application.php(1027): Symfony\Component\Console\Command\Command->run(Object(Drush\Symfony\DrushArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
    #7 vendor/symfony/console/Application.php(273): Symfony\Component\Console\Application->doRunCommand(Object(Consolidation\AnnotatedCommand\AnnotatedCommand), Object(Drush\Symfony\DrushArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput)) 
    #8 vendor/symfony/console/Application.php(149): Symfony\Component\Console\Application->doRun(Object(Drush\Symfony\DrushArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
    #9 vendor/drush/drush/src/Runtime/Runtime.php(118): Symfony\Component\Console\Application->run(Object(Drush\Symfony\DrushArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
    #10 vendor/drush/drush/src/Runtime/Runtime.php(48): Drush\Runtime\Runtime->doRun(Array, Object(Symfony\Component\Console\Output\ConsoleOutput))
    #11 vendor/drush/drush/drush.php(72): Drush\Runtime\Runtime->run(Array)
    #12 vendor/drush/drush/drush(4): require('...')
    #13 {main}. 
    Error: Call to a member function getStatus() on bool in vendor/drush/drush/src/Drupal/Commands/core/MigrateRunnerCommands.php on line 438 #0 [internal function]: Drush\Drupal\Commands\core\MigrateRunnerCommands->stop('unexisting_migr...', Array)
    #1 vendor/consolidation/annotated-command/src/CommandProcessor.php(257): call_user_func_array(Array, Array)
    #2 vendor/consolidation/annotated-command/src/CommandProcessor.php(212): Consolidation\AnnotatedCommand\CommandProcessor->runCommandCallback(Array, Object(Consolidation\AnnotatedCommand\CommandData))
    #3 vendor/consolidation/annotated-command/src/CommandProcessor.php(176): Consolidation\AnnotatedCommand\CommandProcessor->validateRunAndAlter(Array, Array, Object(Consolidation\AnnotatedCommand\CommandData))
    #4 vendor/consolidation/annotated-command/src/AnnotatedCommand.php(311): Consolidation\AnnotatedCommand\CommandProcessor->process(Object(Symfony\Component\Console\Output\ConsoleOutput), Array, Array, Object(Consolidation\AnnotatedCommand\CommandData))
    #5 vendor/symfony/console/Command/Command.php(255): Consolidation\AnnotatedCommand\AnnotatedCommand->execute(Object(Drush\Symfony\DrushArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
    #6 vendor/symfony/console/Application.php(1027): Symfony\Component\Console\Command\Command->run(Object(Drush\Symfony\DrushArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
    #7 vendor/symfony/console/Application.php(273): Symfony\Component\Console\Application->doRunCommand(Object(Consolidation\AnnotatedCommand\AnnotatedCommand), Object(Drush\Symfony\DrushArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
    #8 vendor/symfony/console/Application.php(149): Symfony\Component\Console\Application->doRun(Object(Drush\Symfony\DrushArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
    #9 vendor/drush/drush/src/Runtime/Runtime.php(118): Symfony\Component\Console\Application->run(Object(Drush\Symfony\DrushArgvInput), Object(Symfony\Component\Console\Output\ConsoleOutput))
    #10 vendor/drush/drush/src/Runtime/Runtime.php(48): Drush\Runtime\Runtime->doRun(Array, Object(Symfony\Component\Console\Output\ConsoleOutput))
    #11 vendor/drush/drush/drush.php(72): Drush\Runtime\Runtime->run(Array)
    #12 vendor/drush/drush/drush(4): require('...')
    #13 {main}
    Error: Call to a member function getStatus() on bool in Drush\Drupal\Commands\core\MigrateRunnerCommands->stop() (line 438 of vendor/drush/drush/src/Drupal/Commands/core/MigrateRunnerCommands.php).
     [warning] Drush command terminated abnormally.

Instead, once the changes are in place, this is the new output.

    $ vendor/bin/drush migrate:stop unexisting_migration
     [error]  Migration unexisting_migration does not exist

weitzman commented:
We should emulate validateConfigName() and validate migration names that are passed as arguments.

marvil07 commented:
@weitzman, thanks for the feedback!

I have changed the approach to use an annotated command hook for the validation of the migration name.

marvil07 commented:
I also noticed the same problem for both `migrate:messages` and `migrate:fields-source` commands.
And now that the last commit added a validate hook, it is natural to just use it for those other two cases.
I added a couple of changes about it, and also changed the name and description of this pull request.

@weitzman, again, nice suggestion!

marvil07 commented:
And it also makes sense for `migrate:reset-status`.

weitzman commented:
LGTM. Any concerns, @claudiu-cristea ?

claudiu-cristea commented on src/Drupal/Commands/core/MigrateRunnerCommands.php:
```diff
@@ -759,4 +757,28 @@ protected function getSourceIdKeys(MigrateIdMapInterface $idMap): array
     }
+
+    /**
+     * Validate that a migration name is valid.
```
s/Validate/Validates

claudiu-cristea commented on src/Drupal/Commands/core/MigrateRunnerCommands.php:
```diff
@@ -759,4 +757,28 @@ protected function getSourceIdKeys(MigrateIdMapInterface $idMap): array
+     */
+    public function validateMigrationName(CommandData $commandData)
+    {
+        $arg_name = $commandData->annotationData()->get('validate-migration-name', null) ?: 'migrationId';
```
Let's keep the local vars naming consistent as _lowerCamelCase_. Applies to `$migration_name`, `$arg_name`. Also, why not keeping `$migrationId` when we refer to migration ID, to be consistent with the other methods?

claudiu-cristea replied:
The `null` default value can be removed. If missing a default value, an empty string is returned and the whole line work in the same way

claudiu-cristea commented on src/Drupal/Commands/core/MigrateRunnerCommands.php:
```diff
@@ -759,4 +757,28 @@ protected function getSourceIdKeys(MigrateIdMapInterface $idMap): array
+        $arg_name = $commandData->annotationData()->get('validate-migration-name', null) ?: 'migrationId';
+        $migration_name = $commandData->input()->getArgument($arg_name);
+        /** @var \Drupal\migrate\Plugin\MigrationInterface $migration */
+        $migration = $this->getMigrationPluginManager()->createInstance($migration_name);
```
Creating a new instance is more expensive than checking for existing definition: `$this->getMigrationPluginManager()->hasDefinition(...)`?

claudiu-cristea reviewed (changes requested):
Nice improvement and cleanup. I still have some minor remarks

claudiu-cristea commented on src/Drupal/Commands/core/MigrateRunnerCommands.php:
```diff
@@ -759,4 +757,28 @@ protected function getSourceIdKeys(MigrateIdMapInterface $idMap): array
+        /** @var \Drupal\migrate\Plugin\MigrationInterface $migration */
+        $migration = $this->getMigrationPluginManager()->createInstance($migration_name);
+        if (!$migration) {
+            $msg = dt('Migration "@id" does not exist', ['@id' => $migration_name]);
+            return new CommandError($msg);
```
I guess it can go in a single line. But it's a matter of taste...
